core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -383,11 +383,28 @@ protected boolean booleanEquality(SqlCallBinding binding,
   }
 
   /**
-   * CASE and COALESCE type coercion, collect all the branches types including then
-   * operands and else operands to find a common type, then cast the operands to the common type
-   * when needed.
+   * Coerces CASE WHEN, COALESCE and NULLIF statement branches to a unified type.
+   *
+   * @deprecated Use {@link #caseOrEquivalentCoercion} instead.
+   */
+  @Deprecated @Override public boolean caseWhenCoercion(SqlCallBinding callBinding) {
+    return caseOrEquivalentCoercion(callBinding);
+  }
+
+  /**
+   * Coerces CASE WHEN, COALESCE and NULLIF statement branches to a unified type.
    */
-  @Override public boolean caseWhenCoercion(SqlCallBinding callBinding) {
+  @Override public boolean caseOrEquivalentCoercion(SqlCallBinding callBinding) {
+    if (callBinding.getCall().getKind() == SqlKind.COALESCE) {
+      // For sql statement like: `coalesce(a, b, c)`
+      return coalesceCoercion(callBinding);
+    }
+
+    if (callBinding.getCall().getKind() == SqlKind.NULLIF) {
+      // For sql statement like: `nullif(a, b)`
+      return nullifCoercion(callBinding);
+    }
+
```

Comment:
can you also add an assertion here that the remaining call is CASE?

Reply:
for symmetry you may also want to move this code into a separate private function